vms/evm/uptimetracker/uptime_tracker.go

```diff
@@ -0,0 +1,267 @@
+	if u.deactivatedValidators.Contains(nodeID) {
+		return nil
+	}
+
+	return u.manager.Disconnect(nodeID)
+}
+
+// Sync updates the validator set and writes our state. Sync starts tracking
+// uptimes for all active validators the first time it is called.
+func (u *UptimeTracker) Sync(ctx context.Context) error {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	currentValidatorSet, height, err := u.validatorState.GetCurrentValidatorSet(
+		ctx,
+		u.subnetID,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get current validator set: %w", err)
+	}
+
+	// We are behind and need to update our local state
+	if u.height < int(height) {
+		if err := u.update(height, currentValidatorSet); err != nil {
+			return fmt.Errorf("failed to update validator set: %w", err)
+		}
+	}
+
+	// Initialize uptimes if this is the first time Sync has been called
+	if !u.synced {
+		validationIDs := make([]ids.NodeID, 0)
+		validationIDs = append(validationIDs, u.state.getNodeIDs()...)
+
+		if err := u.manager.StartTracking(validationIDs); err != nil {
+			return fmt.Errorf("failed to start tracking validators: %w", err)
+		}
+
+		u.synced = true
+	}
+
+	if err := u.state.writeState(); err != nil {
+		return fmt.Errorf("failed to write state: %w", err)
+	}
+
+	return nil
+}
+
+func (u *UptimeTracker) update(
+	height uint64,
+	currentValidatorSet map[ids.ID]*validators.GetCurrentValidatorOutput,
+) error {
+	currentValidationIDs := u.state.getValidationIDs()
+	newValidators := currentValidatorSet
+
+	for _, validationID := range currentValidationIDs {
+		// This validator is still in the latest update
+		if _, ok := newValidators[validationID]; ok {
+			continue
+		}
+
+		// This validator was removed in the lastest update
+		nodeID, ok := u.state.getNodeID(validationID)
+		if !ok {
+			return fmt.Errorf("failed to fetch validator %s", validationID)
+		}
+
+		if !u.state.deleteValidator(validationID) {
+			return fmt.Errorf("failed to delete validator %s", validationID)
+		}
+
+		u.deactivatedValidators.Remove(nodeID)
+	}
+
+	// Add or update validators
+	for validationID, newValidator := range newValidators {
+		if _, ok := u.state.getValidatorByValidationID(validationID); ok {
+			// Check if there was a status change
+			if !u.state.updateValidator(validationID, newValidator.IsActive) {
+				continue
+			}
+
+			// If there was a status change we need to activate/deactivate the
+			// validator
+			if newValidator.IsActive {
+				// This validator is now active and is treated as online
+				if err := u.activate(newValidator.NodeID); err != nil {
+					return fmt.Errorf(
+						"failed to activate validator %s: %w",
+						newValidator.NodeID,
+						err,
+					)
+				}
+				continue
+			}
+
+			// This validator is no longer active and is treated as offline
+			if err := u.deactivate(newValidator.NodeID); err != nil {
+				return fmt.Errorf(
+					"failed to deactivate validator %s: %w",
+					newValidator.NodeID,
+					err,
+				)
+			}
+		} else {
```

Comment:
I think we could just add a `continue` here to do an early-exit? if/else is usually harder to read than early-exit code.